### PR TITLE
docs: add adr004, module export structure

### PR DESCRIPTION
**Pull request by Rugvip**
Would like to try this structure out and see if it makes sense.

### CODE REVIEW

File: docs/architecture-decisions/adr004-module-export-structure.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# ADR004: Module Export Structure
+
+| Created    | Status |
+| ---------- | ------ |
+| 2020-05-27 | Open   |
+
+## Context
+
+With a growing number of exports of packages like `@backstage/core`, it is becoming more and more difficult answer questions such as
```
Comment (freben) [inline docs/architecture-decisions/adr004-module-export-structure.md:9]: "difficult to answer"

File: docs/architecture-decisions/adr004-module-export-structure.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+or
+
+> What is exported from this directory?
+
+We currently do not use any pattern for how to structure exports. There is a mix of package-level re-exports deep into the directory tree, shallow re-exports for each directory, exports using `*` and explicit lists of each symbol, etc.
+The mix and lack of predictability makes it difficult to reason about the boundaries of a module, and for example knowing whether is is safe to export a symbol in a given file.
+
+## Decision
+
+We will make each exported symbol traceable through index files all the way down to the root of the package, `src/index.ts`. Each index file will only re-export from it's own immediate directory children, and only index files will have re-exports. This gives a file tree similar to this:
```
Comment (freben) [inline docs/architecture-decisions/adr004-module-export-structure.md:22]: "from its own"

**Review by freben — APPROVED** [review]
Nice!

File: docs/architecture-decisions/adr004-module-export-structure.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+          /ComponentX/index.ts
+                     /ComponentX.tsx
+                     /SubComponentY.tsx
+lib/index.ts
+   /UtilityX/index.ts
+            /UtilityX.ts
+            /helper.ts
+```
+
+To check whether for example `SubComponentY` is exported from the package, it should be possible to traverse the index files towards the root, starting at the adjacent one. If there is any index file that doesn't export the previous one, the symbol is not publicly exported. For example, if `components/ComponentX/index.ts` exports `SubComponentY`, but `components/index.ts` does not re-export `./ComponentX`, one should be certain that `SubComponentY` is not exported outside the package. This rule would be broken if for example the root `index.ts` re-exports `./components/ComponentX`
```
Comment (soapraj) [inline docs/architecture-decisions/adr004-module-export-structure.md:36]: `components/index.ts` does not re-export `./ComponentX` -> `components/index.ts` does not re-export `./SubComponentY` ?

or am I understanding this wrong?

**Review by soapraj — APPROVED** [review]

**Pull request merged**